### PR TITLE
Add MIMIR MIA attacks support, add MIA to TOFU

### PR DESCRIPTION
**Pull request by molereddy**
# What does this PR do?
- support 6 of [MIMIR](https://github.com/iamgroot42/mimir)'s MIA attacks: minK, minK ++, ref, zlib, loss, GradNorm
- simplify and standardize MIA evals structure/code
- Add holdout dataset so TOFU allows for MIA based privleak evaluations: created by generating IID examples with the same GPT-4 endpoint and original prompt from TOFU's original creation
- updated documentation
- document source links for all features implemented in `links.md` (addresses https://github.com/locuslab/open-unlearning/issues/70)
- generalise KS-test, rel diff metrics
- add ES and EM metrics
- add seed in train and eval
- updated leaderboard and repro.md files with new evals

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Have you gone through the documentation of adding new [components](../docs/components.md)?
- [x] Did you make sure to update the documentation with your changes? Here are the pointers to documentation
      [documentation guidelines](../README.md#-further-documentation).

### CODE REVIEW

**Comment by molereddy** [comment]
To-dos:
1. Create and upload holdout splits for TOFU. Edit TOFU_MIA dataset cfg.
2. Run all evals again for the new eval logs structure. Update eval log files here and on HuggingFace. 
3. Update TOFU results.md column with privleak values instead of truth ratio.
4. Add note to README asking to re-download logs.
5. Update documentation accordingly

File: README.md [hunk 1]
```diff
@@ -21,12 +21,19 @@
...
 ## 📖 Overview
 
-We provide efficient and streamlined implementations of the TOFU, MUSE unlearning benchmarks while supporting 6 unlearning methods, 3+ datasets, 6+ evaluation metrics, and 7+ LLMs. Each of these can be easily extended to incorporate more variants.
+We provide efficient and streamlined implementations of the TOFU, MUSE unlearning benchmarks while supporting 6 unlearning methods, 3+ datasets, 6+ evaluation metrics, and 6+ LLM architectures. Each of these can be easily extended to incorporate more variants.
 
 We invite the LLM unlearning community to collaborate by adding new benchmarks, unlearning methods, datasets and evaluation metrics here to expand OpenUnlearning's features, gain feedback from wider usage and drive progress in the field.
 
-> ⚠️ **Notice (Updated: February 27, 2025)**  
-> This repository replaces the original TOFU codebase, which can be found at [`github.com/locuslab/tofu`](https://github.com/locuslab/tofu) and isn't maintained anymore.
+### 📢 Updates
```
Comment (Dornavineeth) [inline README.md:30]: Shall we make this scrollable?

File: configs/data/datasets/TOFU_MIA.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    max_length: 512
+TOFU_QA_holdout:
+  access_key: holdout
+  handler: QADataset
+  args:
+    hf_args:
+      # name: "holdout10"
+      # path: "locuslab/TOFU"
+      path: "json"
+      data_files: "/scratch3/workspace/amekala_umass_edu-ulv2/Vineeth/TOFU_V2/tofu_gen/holdout.jsonl"
```
Comment (Dornavineeth) [inline configs/data/datasets/TOFU_MIA.yaml]: update the holdout dataset to hf path.
Make different splits.

File: configs/eval/tofu_metrics/forget_quality.yaml [hunk 3]
```diff
@@ -13,4 +13,4 @@ pre_compute:
   forget_truth_ratio:
     access_key: forget
 
-handler: forget_quality
\ No newline at end of file
+handler: ks_test
```
Comment (Dornavineeth) [inline configs/eval/tofu_metrics/forget_quality.yaml:16]: Check if we use forget_quality handler as example in docs.

File: configs/eval/tofu_metrics/mia_reference.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# @package eval.tofu.metrics.mia_reference
+defaults:
+  - ../../data/datasets@datasets: TOFU_MIA
+  - ../../collator@collators: DataCollatorForSupervisedDatasetwithIndex
+batch_size: 32
+handler: mia_reference
+reference_model_path: open-unlearning/tofu_Llama-3.2-1B-Instruct_retain90
```
Comment (Dornavineeth) [inline configs/eval/tofu_metrics/mia_reference.yaml]: data split and model name is hardcoded. 
Need to discuss.

File: docs/results.md [hunk 5]
```diff
@@ -20,11 +20,13 @@ For all the experiments below, we used the following setup
...
+
 
 __Note:__ 
-1. Results may vary even with the same effective hyperparameters when trained with modifications to the distributed training setup, including when training on a single GPU. For example: methods such as SimNPO & RMU can be significantly improved with careful tuning. **Please use these numbers only for reproducibility purposes**.
-2. NPO in MUSE: for NPO, the MUSE implementation is inconsistent with the [original paper](https://github.com/licong-lin/negative-preference-optimization) as discussed [here]( https://github.com/jaechan-repo/muse_bench/issues/2). This inconsistency is carried over into implementations like [SimNPO](https://github.com/OPTML-Group/Unlearn-Simple/issues/5). Here, we use the original NPO implementation with the same loss function expression across datasets.
+1. The results in the next section: display only some important subsets of metrics for each benchmark, for more evals see `muse*/*_SUMMARY.json`, `tofu*/evals*/*_SUMMARY.json` files on the [HuggingFace space](https://huggingface.co/datasets/open-unlearning/eval).
```
Comment (Dornavineeth) [inline docs/results.md]: We don't upload metrics for the unlearned model anywhere.

File: todos.txt [hunk 6]
```diff
@@ -0,0 +1,73 @@
+can link major survey papers and awesome-llm-unlearning as well
```
Comment (Dornavineeth) [inline todos.txt]: Delete this file

File: src/evals/metrics/mia/__init__.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+from evals.metrics.mia.min_k_plus_plus import MinKPlusPlusAttack
+from evals.metrics.mia.gradnorm import GradNormAttack
+from evals.metrics.mia.zlib import ZLIBAttack
+from evals.metrics.mia.reference import ReferenceAttack
+
+from evals.metrics.mia.utils import mia_auc
+import logging
+logger = logging.getLogger("metrics")
+
+## NOTE: all MIA attack statistics are signed with -1 as req in order to
```
Comment (Dornavineeth) [inline src/evals/metrics/mia/__init__.py]: Remove comment

File: src/evals/metrics/mia/gradnorm.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+"""
+    Gradient-norm attack. Proposed for MIA in multiple settings, and particularly 
+    experimented for pre-training data and LLMs in https://arxiv.org/abs/2402.17012
+"""
+import torch
+from evals.metrics.mia.all_attacks import Attack
+from evals.metrics.utils import tokenwise_logprobs
+
+class GradNormAttack(Attack):
```
Comment (Dornavineeth) [inline src/evals/metrics/mia/gradnorm.py:13]: Can this be used during training?

File: src/evals/metrics/mia/reference.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+"""
+    Reference-based attacks.
+"""
+from evals.metrics.mia.all_attacks import Attack
+from evals.metrics.utils import evaluate_probability
+
+class ReferenceAttack(Attack):
```
Comment (Dornavineeth) [inline src/evals/metrics/mia/reference.py:9]: Shall we remove this attack?
Because loading a reference model by simply using Automodel can cause issues when we load model with peft, quantization etc.

How about computing this through retain_logs?

File: docs/results.md (hunk 5 above)
Comment (molereddy) [reply docs/results.md]: making this just about examples of metrics

File: configs/eval/tofu_metrics/forget_quality.yaml (hunk 3 above)
Comment (molereddy) [reply configs/eval/tofu_metrics/forget_quality.yaml:16]: checked and fixed

File: README.md (hunk 1 above)
Comment (molereddy) [reply README.md:30]: done

**Comment by molereddy** [comment]
Need to modify date in README before merging

File: README.md [hunk 10]
```diff
@@ -25,14 +25,32 @@ We provide efficient and streamlined implementations of the TOFU, MUSE unlearnin
 
 We invite the LLM unlearning community to collaborate by adding new benchmarks, unlearning methods, datasets and evaluation metrics here to expand OpenUnlearning's features, gain feedback from wider usage and drive progress in the field.
 
+---
+
 ### 📢 Updates
 
-#### [Mar 27, 2025]  
+#### [Mar 28, 2025]
+- **MIA**: Added 6 membership inference attacks—LOSS, ZLib, Reference, GradNorm, MinK, and MinK++ as options in the evaluation metrics.
```
Comment (molereddy) [inline README.md]: Need to modify date in README before merging

**Comment by molereddy** [comment]
Todo: create eval logs for privleak by evaluating retain models. Upload to huggingface and ensure `setup_data.py` populates it

edit: addressed in https://huggingface.co/datasets/open-unlearning/eval/discussions/1. No change required here.

File: .gitignore [hunk 11]
```diff
@@ -1,4 +1,5 @@
 # custom .gitignore
+todos.txt
```
Comment (Dornavineeth) [inline .gitignore]: Remove this.

**Review by Dornavineeth — APPROVED** [review]
